### ISSUE TITLE
Selenium Hub does not update Node's configuration if started after the node

### ISSUE THREAD

**Issue by dev-e** [SeleniumHQ/selenium#5908]
Selenium Version:  
3.12.0
Browser:  
Any
## Expected Behavior -
Node configuration is updated with corresponding Hub parameters
## Actual Behavior -
Configuration for the node is not updated with Hub parameters

![image](https://user-images.githubusercontent.com/22076937/40186754-8dc0d1d4-59fe-11e8-973a-19e24607d32a.png)

```
Config for the hub :
browserTimeout : 240
debug : true
help : false
jettyMaxThreads : -1
port : 4444
role : hub
timeout : 15
enablePassThrough : false
cleanUpCycle : 5000
host : 10.64.214.26
maxSession : 5
hubConfig : /opt/selenium/config.json
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry
Config details :
hub launched with : -browserTimeout 240 -debug true -help false -jettyMaxThreads -1 -port 4444 -role hub -timeout 15 -enablePassThrough false -cleanUpCycle 5000 -host 10.64.214.26 -maxSession 5 -hubConfig /opt/selenium/config.json -capabilityMatcher org.openqa.grid.internal.utils.DefaultCapabilityMatcher -newSessionWaitTimeout -1 -throwOnCapabilityNotPresent true -registry org.openqa.grid.internal.DefaultGridRegistry
the final configuration comes from :
the default :
browserTimeout : 0
debug : false
help : false
port : 4444
role : hub
timeout : 1800
enablePassThrough : true
cleanUpCycle : 5000
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry

updated with params :
browserTimeout : 240
debug : false
help : false
jettyMaxThreads : -1
port : 4444
role : hub
timeout : 15
enablePassThrough : true
cleanUpCycle : 5000
maxSession : 5
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry
```

## Steps to reproduce -
1. Start a node with default parameters, check that it cannot register to the hub
2. Start a hub with environment GRID_TIMEOUT=15
3. Check node configuration after the registration on the hub

**Comment by barancev** [comment]
Please retest with the latest version that is 2.13 at the moment.

**Comment by dev-e** [comment]
@barancev 2.13???

**Comment by barancev** [comment]
3.12, sorry :)

**Comment by dev-e** [comment]
The same sesult
1. Start node
 ```
java -jar selenium-server-standalone-3.12.0.jar -role node -hub http://localhost:4444/grid/register
18:51:56.796 INFO [GridLauncherV3.launch] - Selenium build info: version: '3.12.0', revision: '7c6e0b3'
18:51:56.811 INFO [GridLauncherV3$3.launch] - Launching a Selenium Grid node on port 30873
2018-05-17 18:51:56.959:INFO::main: Logging initialized @658ms to org.seleniumhq.jetty9.util.log.StdErrLog
18:51:57.207 INFO [SeleniumServer.boot] - Selenium Server is up and running on port 30873
18:51:57.208 INFO [GridLauncherV3$3.launch] - Selenium Grid node is up and ready to register to the hub
18:51:57.217 INFO [SelfRegisteringRemote$1.run] - Starting auto registration thread. Will try to register every 5000 ms.
18:51:57.218 INFO [SelfRegisteringRemote.registerToHub] - Registering the node to the hub: http://localhost:4444/grid/register
18:51:57.509 INFO [SelfRegisteringRemote$1.run] - Couldn't register this node: Error sending the registration request: Failed to connect to localhost/127.0.0.1:4444
18:52:02.526 INFO [SelfRegisteringRemote$1.run] - Couldn't register this node: The hub is down or not responding: Failed to connect to localhost/0:0:0:0:0:0:0:1:4444
18:52:07.537 INFO [SelfRegisteringRemote$1.run] - Couldn't register this node: The hub is down or not responding: Failed to connect to localhost/127.0.0.1:4444
```
2. Start hub
```
java -jar selenium-server-standalone-3.12.0.jar -role hub -timeout 15
18:52:32.763 INFO [GridLauncherV3.launch] - Selenium build info: version: '3.12.0', revision: '7c6e0b3'
18:52:32.766 INFO [GridLauncherV3$2.launch] - Launching Selenium Grid hub on port 4444
2018-05-17 18:52:33.110:INFO::main: Logging initialized @744ms to org.seleniumhq.jetty9.util.log.StdErrLog
18:52:33.332 INFO [Hub.start] - Selenium Grid hub is up and running
18:52:33.333 INFO [Hub.start] - Nodes should register to http://172.17.0.1:4444/grid/register/
18:52:33.334 INFO [Hub.start] - Clients should connect to http://172.17.0.1:4444/wd/hub
18:52:38.038 INFO [DefaultGridRegistry.add] - Registered a node http://172.17.0.1:30873
```
3. Check configuration
![image](https://user-images.githubusercontent.com/22076937/40188888-a1d83c48-5a03-11e8-9d65-5f4b81a45d68.png)
```
Config for the hub :
browserTimeout : 0
debug : false
host : 172.17.0.1
port : 4444
role : hub
timeout : 15
cleanUpCycle : 5000
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry
Config details :
hub launched with : -browserTimeout 0 -debug false -host 172.17.0.1 -port 4444 -role hub -timeout 15 -cleanUpCycle 5000 -capabilityMatcher org.openqa.grid.internal.utils.DefaultCapabilityMatcher -newSessionWaitTimeout -1 -throwOnCapabilityNotPresent true -registry org.openqa.grid.internal.DefaultGridRegistry
the final configuration comes from :
the default :
browserTimeout : 0
debug : false
port : 4444
role : hub
timeout : 1800
cleanUpCycle : 5000
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry

updated with params :
browserTimeout : 0
debug : false
port : 4444
role : hub
timeout : 15
cleanUpCycle : 5000
capabilityMatcher : org.openqa.grid.internal.utils.DefaultCapabilityMatcher
newSessionWaitTimeout : -1
throwOnCapabilityNotPresent : true
registry : org.openqa.grid.internal.DefaultGridRegistry
```

**Comment by Niceplace** [comment]
@barancev Also experiencing this issue with 3.12, reproductible scenario (with docker) in the issue just above : https://github.com/SeleniumHQ/docker-selenium/issues/727

**Comment by derinbay** [comment]
@Niceplace When grid hub is starting first, then nodes are creating and connection to grid, this problem occures.
For a workaround, first create the nodes, then create the grid hub. It respects the timeout values like this.

@barancev I hope this will fix soon.

**Comment by barancev** [comment]
I'm not sure we'll be able to fix it in 3.13, I target to 3.14.

**Comment by dev-e** [comment]
@derinbay workaround is vice versa: first start a hub, then start nodes.

**Comment by derinbay** [comment]
@dev-e In my case it's working if I start nodes first. I'm using docker-selenium btw

**Comment by dev-e** [comment]
@derinbay well, I also use selenium in docker from official repo on docker hub. I my case I should run hub first.

**Comment by Niceplace** [comment]
@dev-e @derinbay Introducing a wait between grid / node startup did not fix it for me. I should retry just to be sure but at the time of writing this post, I remember starting the grid and waiting up to 1 minute before starting the nodes with no result.

I'll wait for 3.14 and try again !

**Comment by derinbay** [comment]
Well, after I comment on this post, I tried to add timeout parameters to nodes with SE_OPTS, and it works!

https://gist.github.com/derinbay/ebe8192891950f7edc355de5b8e16030

**Comment by Niceplace** [comment]
Oh, that's not the way I did it at all, cool ! I'll try this.

**Comment by barancev** [comment]
I suppose a node does actually update timeouts from the hub, it's a hub console issue, it shows wrong info about timeouts.

**Comment by Niceplace** [comment]
@barancev are you saying this is purely a UI display issue ? How could we validate that on our end ?

**Comment by barancev** [comment]
In version 3.14 a node should write to stdout information about its configuration, it's more reliable info source than the current UI.

**Comment by diemol** [comment]
Hi,

This is a duplicate of #3064

@barancev is right, it is an issue in the Grid console, showing the wrong values in the Node for timeout and browserTimeout. I explain it in more detail in this [comment](https://github.com/SeleniumHQ/selenium/issues/3064#issuecomment-417959009).

It has been fixed through https://github.com/SeleniumHQ/selenium/commit/88340f45b4038e8cfd8b29f2ccbb9db3ef91b84b and it will be part of the next release.

In the meantime, to check that it is really a issue in the console, you can start the node in debug mode and see in the log where the node is getting the hub values.
